Support Python 3.11

## What was wrong?

Doesn't install on Python 3.11 from master branch

### Code that produced the error

```
pip install .
```

### Full error output 

```sh
(cpy311b1_1) (base) oberstet@intel-nuci7:~/scm/crossbario/eth-account$ pip install .
Processing /home/oberstet/scm/crossbario/eth-account
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting eth-keys<0.5,>=0.4.0
  Using cached eth_keys-0.4.0-py3-none-any.whl (21 kB)
Collecting bitarray<1.3.0,>=1.2.1
  Using cached bitarray-1.2.2.tar.gz (48 kB)
  Preparing metadata (setup.py) ... done
Requirement already satisfied: hexbytes<1,>=0.1.0 in /home/oberstet/cpy311b1_1/lib/python3.11/site-packages (from eth-account==0.6.1) (0.2.2)
Requirement already satisfied: eth-utils<3,>=2.0.0 in /home/oberstet/cpy311b1_1/lib/python3.11/site-packages (from eth-account==0.6.1) (2.0.0)
Collecting eth-keyfile<0.7.0,>=0.6.0
  Using cached eth_keyfile-0.6.0-py3-none-any.whl (6.6 kB)
Requirement already satisfied: eth-abi<4,>=3.0.0 in /home/oberstet/cpy311b1_1/lib/python3.11/site-packages (from eth-account==0.6.1) (3.0.0)
Collecting eth-rlp<1,>=0.3.0
  Using cached eth_rlp-0.3.0-py3-none-any.whl (5.0 kB)
Collecting rlp<4,>=1.0.0
  Using cached rlp-3.0.0-py2.py3-none-any.whl (20 kB)
Requirement already satisfied: eth-typing<4.0.0,>=3.0.0 in /home/oberstet/cpy311b1_1/lib/python3.11/site-packages (from eth-abi<4,>=3.0.0->eth-account==0.6.1) (3.1.0)
Requirement already satisfied: parsimonious<0.9.0,>=0.8.0 in /home/oberstet/cpy311b1_1/lib/python3.11/site-packages (from eth-abi<4,>=3.0.0->eth-account==0.6.1) (0.8.1)
Requirement already satisfied: pycryptodome<4,>=3.6.6 in /home/oberstet/cpy311b1_1/lib/python3.11/site-packages (from eth-keyfile<0.7.0,>=0.6.0->eth-account==0.6.1) (3.14.1)
Requirement already satisfied: eth-hash<0.4.0,>=0.3.1 in /home/oberstet/cpy311b1_1/lib/python3.11/site-packages (from eth-utils<3,>=2.0.0->eth-account==0.6.1) (0.3.2)
Requirement already satisfied: cytoolz<1.0.0,>=0.10.1 in /home/oberstet/cpy311b1_1/lib/python3.11/site-packages (from eth-utils<3,>=2.0.0->eth-account==0.6.1) (0.11.2)
Requirement already satisfied: toolz>=0.8.0 in /home/oberstet/cpy311b1_1/lib/python3.11/site-packages (from cytoolz<1.0.0,>=0.10.1->eth-utils<3,>=2.0.0->eth-account==0.6.1) (0.11.2)
Requirement already satisfied: six>=1.9.0 in /home/oberstet/cpy311b1_1/lib/python3.11/site-packages (from parsimonious<0.9.0,>=0.8.0->eth-abi<4,>=3.0.0->eth-account==0.6.1) (1.16.0)
Building wheels for collected packages: eth-account, bitarray
  Building wheel for eth-account (pyproject.toml) ... done
  Created wheel for eth-account: filename=eth_account-0.6.1-py3-none-any.whl size=103555 sha256=39dcb982e910105a1331e434fc16fbcf0ca87a664fa20af26cceb7877eb78299
  Stored in directory: /home/oberstet/.cache/pip/wheels/e3/67/b2/f6d2b1992ec550947fb06293545e4a6f8d9ff653e001bb68bf
  Building wheel for bitarray (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [58 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-3.11
      creating build/lib.linux-x86_64-3.11/bitarray
      copying bitarray/__init__.py -> build/lib.linux-x86_64-3.11/bitarray
      copying bitarray/test_util.py -> build/lib.linux-x86_64-3.11/bitarray
      copying bitarray/util.py -> build/lib.linux-x86_64-3.11/bitarray
      copying bitarray/test_bitarray.py -> build/lib.linux-x86_64-3.11/bitarray
      running build_ext
      building 'bitarray._bitarray' extension
      creating build/temp.linux-x86_64-3.11
      creating build/temp.linux-x86_64-3.11/bitarray
      gcc -pthread -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/home/oberstet/cpy311b1_1/include -I/home/oberstet/cpy311b1/include/python3.11 -c bitarray/_bitarray.c -o build/temp.linux-x86_64-3.11/bitarray/_bitarray.o
      bitarray/_bitarray.c: In function ‘resize’:
      bitarray/_bitarray.c:126:23: error: lvalue required as left operand of assignment
        126 |         Py_SIZE(self) = newsize;
            |                       ^
      bitarray/_bitarray.c:151:19: error: lvalue required as left operand of assignment
        151 |     Py_SIZE(self) = newsize;
            |                   ^
      bitarray/_bitarray.c: In function ‘newbitarrayobject’:
      bitarray/_bitarray.c:172:18: error: lvalue required as left operand of assignment
        172 |     Py_SIZE(obj) = nbytes;
            |                  ^
      bitarray/_bitarray.c: In function ‘bitarray_fromfile’:
      bitarray/_bitarray.c:1429:5: warning: ‘PyEval_CallObjectWithKeywords’ is deprecated [-Wdeprecated-declarations]
       1429 |     result = PyEval_CallObject(reader, rargs);
            |     ^~~~~~
      In file included from /home/oberstet/cpy311b1/include/python3.11/Python.h:95,
                       from bitarray/_bitarray.c:12:
      /home/oberstet/cpy311b1/include/python3.11/ceval.h:27:43: note: declared here
         27 | Py_DEPRECATED(3.9) PyAPI_FUNC(PyObject *) PyEval_CallObjectWithKeywords(
            |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      bitarray/_bitarray.c: In function ‘bitarray_tofile’:
      bitarray/_bitarray.c:1568:5: warning: ‘PyEval_CallObjectWithKeywords’ is deprecated [-Wdeprecated-declarations]
       1568 |     result = PyEval_CallObject(writer, args);
            |     ^~~~~~
      In file included from /home/oberstet/cpy311b1/include/python3.11/Python.h:95,
                       from bitarray/_bitarray.c:12:
      /home/oberstet/cpy311b1/include/python3.11/ceval.h:27:43: note: declared here
         27 | Py_DEPRECATED(3.9) PyAPI_FUNC(PyObject *) PyEval_CallObjectWithKeywords(
            |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      bitarray/_bitarray.c: In function ‘PyInit__bitarray’:
      bitarray/_bitarray.c:3273:28: error: lvalue required as left operand of assignment
       3273 |     Py_TYPE(&Bitarraytype) = &PyType_Type;
            |                            ^
      bitarray/_bitarray.c:3274:31: error: lvalue required as left operand of assignment
       3274 |     Py_TYPE(&SearchIter_Type) = &PyType_Type;
            |                               ^
      bitarray/_bitarray.c:3275:31: error: lvalue required as left operand of assignment
       3275 |     Py_TYPE(&DecodeIter_Type) = &PyType_Type;
            |                               ^
      bitarray/_bitarray.c:3276:33: error: lvalue required as left operand of assignment
       3276 |     Py_TYPE(&BitarrayIter_Type) = &PyType_Type;
            |                                 ^
      error: command '/usr/bin/gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for bitarray
  Running setup.py clean for bitarray
Successfully built eth-account
Failed to build bitarray
Installing collected packages: bitarray, rlp, eth-keys, eth-rlp, eth-keyfile, eth-account
  Attempting uninstall: bitarray
    Found existing installation: bitarray 2.6.0
    Uninstalling bitarray-2.6.0:
      Successfully uninstalled bitarray-2.6.0
  Running setup.py install for bitarray ... error
  error: subprocess-exited-with-error
  
  × Running setup.py install for bitarray did not run successfully.
  │ exit code: 1
  ╰─> [58 lines of output]
      running install
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-3.11
      creating build/lib.linux-x86_64-3.11/bitarray
      copying bitarray/__init__.py -> build/lib.linux-x86_64-3.11/bitarray
      copying bitarray/test_util.py -> build/lib.linux-x86_64-3.11/bitarray
      copying bitarray/util.py -> build/lib.linux-x86_64-3.11/bitarray
      copying bitarray/test_bitarray.py -> build/lib.linux-x86_64-3.11/bitarray
      running build_ext
      building 'bitarray._bitarray' extension
      creating build/temp.linux-x86_64-3.11
      creating build/temp.linux-x86_64-3.11/bitarray
      gcc -pthread -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/home/oberstet/cpy311b1_1/include -I/home/oberstet/cpy311b1/include/python3.11 -c bitarray/_bitarray.c -o build/temp.linux-x86_64-3.11/bitarray/_bitarray.o
      bitarray/_bitarray.c: In function ‘resize’:
      bitarray/_bitarray.c:126:23: error: lvalue required as left operand of assignment
        126 |         Py_SIZE(self) = newsize;
            |                       ^
      bitarray/_bitarray.c:151:19: error: lvalue required as left operand of assignment
        151 |     Py_SIZE(self) = newsize;
            |                   ^
      bitarray/_bitarray.c: In function ‘newbitarrayobject’:
      bitarray/_bitarray.c:172:18: error: lvalue required as left operand of assignment
        172 |     Py_SIZE(obj) = nbytes;
            |                  ^
      bitarray/_bitarray.c: In function ‘bitarray_fromfile’:
      bitarray/_bitarray.c:1429:5: warning: ‘PyEval_CallObjectWithKeywords’ is deprecated [-Wdeprecated-declarations]
       1429 |     result = PyEval_CallObject(reader, rargs);
            |     ^~~~~~
      In file included from /home/oberstet/cpy311b1/include/python3.11/Python.h:95,
                       from bitarray/_bitarray.c:12:
      /home/oberstet/cpy311b1/include/python3.11/ceval.h:27:43: note: declared here
         27 | Py_DEPRECATED(3.9) PyAPI_FUNC(PyObject *) PyEval_CallObjectWithKeywords(
            |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      bitarray/_bitarray.c: In function ‘bitarray_tofile’:
      bitarray/_bitarray.c:1568:5: warning: ‘PyEval_CallObjectWithKeywords’ is deprecated [-Wdeprecated-declarations]
       1568 |     result = PyEval_CallObject(writer, args);
            |     ^~~~~~
      In file included from /home/oberstet/cpy311b1/include/python3.11/Python.h:95,
                       from bitarray/_bitarray.c:12:
      /home/oberstet/cpy311b1/include/python3.11/ceval.h:27:43: note: declared here
         27 | Py_DEPRECATED(3.9) PyAPI_FUNC(PyObject *) PyEval_CallObjectWithKeywords(
            |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      bitarray/_bitarray.c: In function ‘PyInit__bitarray’:
      bitarray/_bitarray.c:3273:28: error: lvalue required as left operand of assignment
       3273 |     Py_TYPE(&Bitarraytype) = &PyType_Type;
            |                            ^
      bitarray/_bitarray.c:3274:31: error: lvalue required as left operand of assignment
       3274 |     Py_TYPE(&SearchIter_Type) = &PyType_Type;
            |                               ^
      bitarray/_bitarray.c:3275:31: error: lvalue required as left operand of assignment
       3275 |     Py_TYPE(&DecodeIter_Type) = &PyType_Type;
            |                               ^
      bitarray/_bitarray.c:3276:33: error: lvalue required as left operand of assignment
       3276 |     Py_TYPE(&BitarrayIter_Type) = &PyType_Type;
            |                                 ^
      error: command '/usr/bin/gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  WARNING: No metadata found in /home/oberstet/cpy311b1_1/lib/python3.11/site-packages
  Rolling back uninstall of bitarray
  Moving to /home/oberstet/cpy311b1_1/lib/python3.11/site-packages/bitarray-2.6.0.dist-info/
   from /home/oberstet/cpy311b1_1/lib/python3.11/site-packages/~itarray-2.6.0.dist-info
  Moving to /home/oberstet/cpy311b1_1/lib/python3.11/site-packages/bitarray/
   from /home/oberstet/cpy311b1_1/lib/python3.11/site-packages/~itarray
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> bitarray

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.
WARNING: You are using pip version 22.0.4; however, version 22.1 is available.
You should consider upgrading via the '/home/oberstet/cpy311b1_1/bin/python3 -m pip install --upgrade pip' command.
```

### Environment

Python 3.11.0b1 on Linux

## How can it be fixed?

Adjust version dependency limits:

```
(cpy311b1_1) (base) oberstet@intel-nuci7:~/scm/crossbario/eth-account$ git diff
diff --git a/setup.py b/setup.py
index 96f333e..7955bfb 100644
--- a/setup.py
+++ b/setup.py
@@ -61,7 +61,7 @@ setup(
         "hdaccount/wordlist/*.txt",
     ]},
     install_requires=[
-        "bitarray>=1.2.1,<1.3.0",
+        "bitarray>=2.6.0,<3",
         "eth-abi>=3.0.0,<4",
         "eth-keyfile>=0.6.0,<0.7.0",
         "eth-keys>=0.4.0,<0.5",
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -130,6 +130,18 @@ jobs:
       - image: cimg/python:3.10
         environment:
           TOXENV: py310-core
+  py311-lint:
+    <<: *common
+    docker:
+      - image: cimg/python:3.11
+        environment:
+          TOXENV: py311-lint
+  py311-core:
+    <<: *common
+    docker:
+      - image: cimg/python:3.11
+        environment:
+          TOXENV: py311-core
   py37-integration:
     <<: *integration
     docker:
@@ -154,6 +166,12 @@ jobs:
       - image: cimg/python:3.10
         environment:
           TOXENV: py310-integration
+  py311-integration:
+    <<: *integration
+    docker:
+      - image: cimg/python:3.11
+        environment:
+          TOXENV: py311-integration
   py37-wheel-cli:
     <<: *common
     docker:
@@ -178,6 +196,12 @@ jobs:
       - image: cimg/python:3.10
         environment:
           TOXENV: py310-wheel-cli
+  py311-wheel-cli:
+    <<: *common
+    docker:
+      - image: cimg/python:3.11
+        environment:
+          TOXENV: py311-wheel-cli
 workflows:
   version: 2
   test:
@@ -187,15 +211,19 @@ workflows:
       - py38-lint
       - py39-lint
       - py310-lint
+      - py311-lint
       - py37-core
       - py38-core
       - py39-core
       - py310-core
+      - py311-core
       - py37-integration
       - py38-integration
       - py39-integration
       - py310-integration
+      - py311-integration
       - py37-wheel-cli
       - py38-wheel-cli
       - py39-wheel-cli
       - py310-wheel-cli
+      - py311-wheel-cli

--- a/newsfragments/212.feature.rst
+++ b/newsfragments/212.feature.rst
@@ -1,0 +1,1 @@
+Add support for Python 3.11

--- a/setup.py
+++ b/setup.py
@@ -66,7 +66,7 @@ setup(
     },
     install_requires=[
         "bitarray>=2.4.0,<3",
-        "eth-abi>=3.0.1",
+        "eth-abi>=4.0.0-b.2",
         "eth-keyfile>=0.6.0,<0.7.0",
         "eth-keys>=0.4.0,<0.5",
         "eth-rlp>=0.3.0,<1",

--- a/tox.ini
+++ b/tox.ini
@@ -48,7 +48,7 @@ commands=
     black {toxinidir}/eth_account {toxinidir}/tests --check
     pydocstyle {toxinidir}/eth_account {toxinidir}/tests
 
-[testenv:py{37,38,39,310}-lint]
+[testenv:py{37,38,39,310,311}-lint]
 extras={[common-lint]extras}
 commands={[common-lint]commands}
 
@@ -68,31 +68,7 @@ commands=
     /bin/bash -c 'pip install --upgrade "$(ls dist/eth_account-*-py3-none-any.whl)" --progress-bar off'
     python -c "from eth_account import Account"
 
-[testenv:py37-wheel-cli]
-deps={[common-wheel-cli]deps}
-allowlist_externals={[common-wheel-cli]allowlist_externals}
-commands={[common-wheel-cli]commands}
-skip_install=true
-
-[testenv:py38-wheel-cli]
-deps={[common-wheel-cli]deps}
-allowlist_externals={[common-wheel-cli]allowlist_externals}
-commands={[common-wheel-cli]commands}
-skip_install=true
-
-[testenv:py39-wheel-cli]
-deps={[common-wheel-cli]deps}
-allowlist_externals={[common-wheel-cli]allowlist_externals}
-commands={[common-wheel-cli]commands}
-skip_install=true
-
-[testenv:py310-wheel-cli]
-deps={[common-wheel-cli]deps}
-allowlist_externals={[common-wheel-cli]allowlist_externals}
-commands={[common-wheel-cli]commands}
-skip_install=true
-
-[testenv:py311-wheel-cli]
+[testenv:py{37,38,39,310,311}-wheel-cli]
 deps={[common-wheel-cli]deps}
 allowlist_externals={[common-wheel-cli]allowlist_externals}
 commands={[common-wheel-cli]commands}

--- a/tox.ini
+++ b/tox.ini
@@ -1,10 +1,10 @@
 [tox]
 envlist=
-    py{37,38,39,310}-core
-    py{37,38,39,310}-integration
-    py{37,38,39,310}-lint
+    py{37,38,39,310,311}-core
+    py{37,38,39,310,311}-integration
+    py{37,38,39,310,311}-lint
     docs
-    py{37,38,39,310}-wheel-cli
+    py{37,38,39,310,311}-wheel-cli
 
 [isort]
 combine_as_imports=True
@@ -33,6 +33,7 @@ basepython =
     py38: python3.8
     py39: python3.9
     py310: python3.10
+    py311: python3.11
 extras=
     test
     docs: doc
@@ -86,6 +87,12 @@ commands={[common-wheel-cli]commands}
 skip_install=true
 
 [testenv:py310-wheel-cli]
+deps={[common-wheel-cli]deps}
+allowlist_externals={[common-wheel-cli]allowlist_externals}
+commands={[common-wheel-cli]commands}
+skip_install=true
+
+[testenv:py311-wheel-cli]
 deps={[common-wheel-cli]deps}
 allowlist_externals={[common-wheel-cli]allowlist_externals}
 commands={[common-wheel-cli]commands}

